### PR TITLE
PHOENIX-6142: Make DDL operations resilient to orphan parent->child linking rows in SYSTEM.CHILD_LINK

### PR DESCRIPTION
**Pull request by ChinmaySKulkarni**

### CODE REVIEW

**Comment by ChinmaySKulkarni** [comment]
Please review @yanxinyi @jpisaac @gjacoby126 @kadirozde @twdsilva

**Comment by twdsilva** [comment]
+1, nice work thanks for fixing all these scenarios.

**Comment by ChinmaySKulkarni** [comment]
Thanks for the review @twdsilva! 
@kadirozde @gjacoby126 @jpisaac @yanxinyi as discussed offline, I have written a design doc outlining my approach: [PHOENIX-6142 Design Doc](https://docs.google.com/document/d/1ZWysagBG84yapT6bIyeKljC-2ydQSgtQ63jl7eiZnFs/edit#). Please take a look. It would be great to get your reviews as well. The push I did after Thomas's review does not contain any code changes. They are just extra comments.

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 32s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ 4.x Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  32m  6s |  4.x passed  |
| +1 :green_heart: |  compile  |   1m  3s |  4.x passed  |
| +1 :green_heart: |  checkstyle  |   2m 30s |  4.x passed  |
| +1 :green_heart: |  javadoc  |   0m 51s |  4.x passed  |
| +0 :ok: |  spotbugs  |   3m 23s |  phoenix-core in 4.x has 956 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| -1 :x: |  mvninstall  |   1m  2s |  root in the patch failed.  |
| -1 :x: |  compile  |   0m 53s |  phoenix-core in the patch failed.  |
| -1 :x: |  javac  |   0m 53s |  phoenix-core in the patch failed.  |
| -1 :x: |  checkstyle  |   2m 37s |  phoenix-core: The patch generated 459 new + 5762 unchanged - 387 fixed = 6221 total (was 6149)  |
| -1 :x: |  whitespace  |   0m  0s |  The patch 1 line(s) with tabs.  |
| -1 :x: |  javadoc  |   0m 47s |  phoenix-core generated 1 new + 99 unchanged - 1 fixed = 100 total (was 100)  |
| -1 :x: |  spotbugs  |   0m 48s |  phoenix-core in the patch failed.  |
||| _ Other Tests _ |
| -1 :x: |  unit  |   0m 50s |  phoenix-core in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 10s |  The patch does not generate ASF License warnings.  |
|  |   |  48m 25s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/4/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/903 |
| JIRA Issue | PHOENIX-6142 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaseanti checkstyle compile |
| uname | Linux 724d19548194 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | 4.x / 264310b |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| mvninstall | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/4/artifact/yetus-general-check/output/patch-mvninstall-root.txt |
| compile | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/4/artifact/yetus-general-check/output/patch-compile-phoenix-core.txt |
| javac | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/4/artifact/yetus-general-check/output/patch-compile-phoenix-core.txt |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/4/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| whitespace | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/4/artifact/yetus-general-check/output/whitespace-tabs.txt |
| javadoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/4/artifact/yetus-general-check/output/diff-javadoc-javadoc-phoenix-core.txt |
| spotbugs | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/4/artifact/yetus-general-check/output/patch-spotbugs-phoenix-core.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/4/artifact/yetus-general-check/output/patch-unit-phoenix-core.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/4/testReport/ |
| Max. process+thread count | 94 (vs. ulimit of 30000) |
| modules | C: phoenix-core U: phoenix-core |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/4/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 30s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ 4.x Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  26m 21s |  4.x passed  |
| +1 :green_heart: |  compile  |   0m 54s |  4.x passed  |
| +1 :green_heart: |  checkstyle  |   3m  2s |  4.x passed  |
| +1 :green_heart: |  javadoc  |   0m 43s |  4.x passed  |
| +0 :ok: |  spotbugs  |   2m 50s |  phoenix-core in 4.x has 956 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  22m 49s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 56s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 56s |  the patch passed  |
| -1 :x: |  checkstyle  |   3m  7s |  phoenix-core: The patch generated 417 new + 5804 unchanged - 345 fixed = 6221 total (was 6149)  |
| -1 :x: |  whitespace  |   0m  0s |  The patch 1 line(s) with tabs.  |
| -1 :x: |  javadoc  |   0m 43s |  phoenix-core generated 1 new + 99 unchanged - 1 fixed = 100 total (was 100)  |
| -1 :x: |  spotbugs  |   3m  6s |  phoenix-core generated 1 new + 953 unchanged - 3 fixed = 954 total (was 956)  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 125m 23s |  phoenix-core in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 38s |  The patch does not generate ASF License warnings.  |
|  |   | 193m 48s |   |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:phoenix-core |
|  |  Found reliance on default encoding in org.apache.phoenix.util.ViewUtil.findImmediateRelatedViews(Table, byte[], byte[], byte[], PTable$LinkType, long):in org.apache.phoenix.util.ViewUtil.findImmediateRelatedViews(Table, byte[], byte[], byte[], PTable$LinkType, long): String.getBytes()  At ViewUtil.java:[line 287] |
| Failed junit tests | phoenix.end2end.PhoenixRuntimeIT |
|   | phoenix.end2end.ToCharFunctionIT |
|   | phoenix.end2end.DeleteIT |
|   | phoenix.end2end.IndexToolForNonTxGlobalIndexIT |
|   | phoenix.end2end.UpsertSelectIT |
|   | phoenix.end2end.index.ImmutableIndexExtendedIT |
|   | phoenix.end2end.DropSchemaIT |
|   | phoenix.end2end.index.IndexMaintenanceIT |
|   | phoenix.end2end.OrphanViewToolIT |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/6/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/903 |
| JIRA Issue | PHOENIX-6142 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaseanti checkstyle compile |
| uname | Linux c35418f930ab 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | 4.x / 264310b |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/6/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| whitespace | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/6/artifact/yetus-general-check/output/whitespace-tabs.txt |
| javadoc | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/6/artifact/yetus-general-check/output/diff-javadoc-javadoc-phoenix-core.txt |
| spotbugs | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/6/artifact/yetus-general-check/output/new-spotbugs-phoenix-core.html |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/6/artifact/yetus-general-check/output/patch-unit-phoenix-core.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/6/testReport/ |
| Max. process+thread count | 6414 (vs. ulimit of 30000) |
| modules | C: phoenix-core U: phoenix-core |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-903/6/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by ChinmaySKulkarni** [comment]
None of the tests that failed in the build above are failing locally. 
Re-running the preCommit manually to get another run.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java [hunk 1]
```diff
@@ -294,7 +321,246 @@ public void testRecreateDroppedTableWithChildViews() throws Exception {
...
+                        cqs.getConfiguration(), EMPTY_BYTE_ARRAY, BASE_TABLE_SCHEMA.getBytes(),
+                        parent2TableName.getBytes(), HConstants.LATEST_TIMESTAMP, false);
+                legitChildViews = allDescendants.getFirst();
+                orphanViews = allDescendants.getSecond();
+                // All of the orphan links are legit views of the other parent so they don't count
+                // as orphan views for this parent
+                assertTrue(orphanViews.isEmpty());
+                assertLegitChildViews(expectedLegitChildViewsListForParent2, legitChildViews);
+
+                // Drop the legitimate level 1 view that was on top of parent1
```
Comment (jpisaac) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java:480]: Wondering whether you want to add a test - 
that validates that the order of dropping does not matter, in other words dropping level1view2 first.
Also, wondering whether adding multiple global views(level1 views for a parent) would make for a more robust test suite?

**Review by jpisaac — COMMENTED** [review]
+1 LGTM, except a request for a couple of more tests.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java (hunk 1 above)
Comment (ChinmaySKulkarni) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/ViewMetadataIT.java:480]: Sure, I can add more global views.

**Comment by ChinmaySKulkarni** [comment]
Thanks for the review @jpisaac , I will modify the test and commit after a successful preCommit

**Pull request merged**